src/main/java/sparksoniq/spark/iterator/flowr/ReturnClauseSparkIterator.java

```diff
@@ -49,8 +61,66 @@ public ReturnClauseSparkIterator(RuntimeTupleIterator child, RuntimeIterator exp
         return itemRDD;
     }
 
-    private JavaRDD<Item> itemRDD;
-    
-    private RuntimeTupleIterator _child;
+    @Override
+    protected boolean hasNextLocal() {
+        return _hasNext;
+    }
+
+    @Override
+    protected Item nextLocal() {
+        if(_hasNext == true){
+            Item result = _nextLocalResult;  // save the result to be returned
+            setNextLocalResult();            // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in Object Lookup", getMetadata());
+    }
+
+    @Override
+    protected void openLocal(DynamicContext context) {
+        this._currentDynamicContext = context;
+
+        _child.open(context);
+        setNextLocalResult();
+    }
+
+    private void setNextLocalResult() {
+        _nextLocalResult = null;
+        if (_expression.hasNext()) {
```

Comment:
Here we have to add a boolean field to know whether _expression was opened yet or not. The first time, the control flow must first go to the else close to make sure _expression() gets opened with the first dynamic context of the tuple stream.

Reply:
Given that _expression's state is consistent the hasNext should return false for non-opened iterators. So the control flow will indeed always go to the else first for the first iteration. However, if this feels hacky*, I can add the boolean solution as well.

*: in the first iteration close() is called on a not opened iterator (_expression), however this call has no side effects in the current implementation.

Reply:
Yes -- it is cleaner to not "abuse" hasNext() to test for openness. In the future, we may even want to add an isOpen() method to the iterators that takes care of that boolean, but for now an explicit boolean is fine.